contracts/ZapV2.sol

```diff
@@ -0,0 +1,246 @@
+            ROUTER.swapExactTokensForTokensSupportingFeeOnTransferTokens(
+                amountIn,
+                amountOutMin,
+                path,
+                to,
+                block.timestamp
+            );
+            amountOut = tokenOut.balanceOf(address(this)) - balanceBefore;
+        } else {
+            if (to != address(this)) {
+                IERC20 token = IERC20(path[0]);
+                uint256 balanceBefore = token.balanceOf(to);
+                token.safeTransfer(to, amountIn);
+                amountOut = token.balanceOf(to) - balanceBefore;
+            } else {
+                amountOut = amountIn;
+            }
+        }
+        require(amountOut >= amountOutMin, "ZapV2: INSUFFICIENT_TOKEN_AMOUNT");
+    }
+
+    function _zapInToken(
+        address tokenFrom,
+        uint256 amountFrom,
+        uint256 amount0Min,
+        uint256 amount1Min,
+        address[] calldata pathToToken0,
+        address[] calldata pathToToken1,
+        address to
+    ) private returns (uint256 liquidity) {
+        _approveTokenIfNeeded(tokenFrom, amountFrom);
+        require(pathToToken0[0] == tokenFrom && pathToToken1[0] == tokenFrom, "ZapV2: INVALID_PATH");
+
+        uint256 sellAmount = amountFrom / 2;
+        uint256 amount0 = _swapExactTokensForTokens(sellAmount, 0, pathToToken0, address(this));
+        uint256 amount1 = _swapExactTokensForTokens(amountFrom - sellAmount, 0, pathToToken1, address(this));
+
+        require(amount0 >= amount0Min, "ZapV2: INSUFFICIENT_A_AMOUNT");
+        require(amount1 >= amount1Min, "ZapV2: INSUFFICIENT_B_AMOUNT");
+
+        liquidity = _addLiquidity(amount0, amount1, amount0Min, amount1Min, pathToToken0, pathToToken1, to);
+    }
+
+    function _zapOutToken(
+        address pairFrom,
+        uint256 amountFrom,
+        uint256 amountToMin,
+        address[] calldata path0,
+        address[] calldata path1,
+        address to
+    ) private returns (uint256 balance) {
+        IERC20(pairFrom).safeTransferFrom(msg.sender, address(this), amountFrom);
+        _approveTokenIfNeeded(pairFrom, amountFrom);
+
+        (address[] calldata pathFromToken0, address[] calldata pathFromToken1) = _orderPath(pairFrom, path0, path1);
+
+        {
+            (address token0, address token1) = _getTokens(pathFromToken0, pathFromToken1);
+            require(token0 == token1, "ZapV2: INVALID_PATH");
+        }
+
+        (uint256 balance0, uint256 balance1) = _removeLiquidity(pairFrom, amountFrom);
+
+        balance = _swapExactTokensForTokens(balance0, 0, pathFromToken0, to);
+        balance += _swapExactTokensForTokens(balance1, 0, pathFromToken1, to);
+
+        require(balance >= amountToMin, "ZapV2: INSUFFICIENT_TOKEN_AMOUNT");
+    }
+
+    /* ========== RESTRICTED FUNCTIONS ========== */
+
+    function withdraw(address token) external onlyOwner {
+        if (token == address(0)) {
+            payable(owner()).transfer(address(this).balance);
```

Comment:
question: is `transfer` or `call` the standard? i would maybe suggest using `call`, i.e. `owner().call{value: address(this.balance)}("")`

Reply:
Call is the new standard, I just used the previous `withdraw` from `ZapV1`, updated this